### ISSUE TITLE
Improve type errors

### ISSUE THREAD

**Issue by jeromesimeon** [accordproject/ergo#374]
The type errors are a bit obtuse at the moment, we should try and provide more information to the user.

**Comment by jeromesimeon** [comment]
For constructing or accessing objects:
```
Welcome to ERGOTOP version 0.2.0
ergo$ define asset Home { address : String, bedrooms: Integer }
ergo$ define constant x : Asset = Home{ address: "John St", bedrooms: 1 }
ergo$ define constant x : Event = Home{ address: "John St", bedrooms: 1 }
[TypeError at line 1 col 0] `Constant' type mismatch
define constant x : Event = Home{ address: "John St", bedrooms: 1 }
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ergo$ define constant x : Asset = Home{ address: "John St" }
[TypeError at line 1 col 28] Concept name doesn't match data
define constant x : Asset = Home{ address: "John St" }
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^
ergo$ return x;
Response. ~org.accordproject.ergotop.Home {address: "John St", bedrooms: 1}
  :  {emit: Unit[], response: ~org.hyperledger.composer.system.Asset, state: Unit}?
ergo$ return x.address;
[TypeError at line 1 col 7] Ill-typed unary operation.
return x.address;
       ^^^^^^^^^ 
ergo$ return x.somethingelse;
[TypeError at line 1 col 7] Ill-typed unary operation.
return x.somethingelse;
       ^^^^^^^^^^^^^^^ 
ergo$ 
```

**Comment by jeromesimeon** [comment]
Some things that would be improvements:
- [ ] Better error messages for subtyping checks (on function parameters, type assertions) that prints the expected and effective types
- [ ] Better error messages for accessing a field of an objet (e.g., `field somethingelse not in asset Home` ?
- [ ] Better error messages for constructing objects (e.g., `asset Home should have a field bedrooms` `field bedroom should be of type Integer`)

**Comment by jeromesimeon** [comment]
Fixes in #433
